### PR TITLE
Optimize MSQ realtime queries

### PR DESCRIPTION
**Pull request by adarshsanjeev**
## Major changes

Currently, while reading results from realtime tasks, requests are sent on a segment level. This is slightly wasteful, as when contacting a data servers, it is possible to transfer results for all segments which it is hosting, instead of only one segment at a time. 

One change this PR makes is to group the segments on the basis of servers. This reduces the number of queries to data servers made. Since we don't have access to the number of rows for realtime segments, the grouping is done with a fixed estimated number of rows for each realtime segment.

 If the segments have already been handed off, the earlier approach made was to fetch the segments from deep storage. 

After the above changes, a single worker processor thread might have to fetch all segments from deepstorage, missing an oppurtunity for parellelism, since each segment fetch from deepstorage, is a separate operation. This PR also edits the ChainedProcessorManager. The initial BaseLeafFrameProcessors return the list of segments which were handed off. The ProcessorManager then uses this information to create new list of frame processors, handle the newer handed off segments.

- Optimize MSQ queries to realtime tasks, by grouping segments served by the same server together.
- Modify BaseLeafFrameProcessors and ChainedProcessorManager to take handle handed of segments
- Update DataServerResponseHandler to support backpressure and query timeout.

<hr>

This PR has:

- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/DataServerQueryResult.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+{
+
+  private final List<Yielder<RowType>> resultsYielders;
+
+  private final SegmentsInputSlice segmentsInputSlice;
+
+  public DataServerQueryResult(
+      List<Yielder<RowType>> resultsYielders,
+      List<RichSegmentDescriptor> handedOffSegments,
+      String dataSource
```
Comment (cryptoe) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/DataServerQueryResult.java:44]: The datasource is never set.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/scan/ScanQueryFrameProcessor.java [hunk 2]
```diff
@@ -192,39 +196,45 @@ private static ScanQuery prepareScanQueryForDataServer(@NotNull ScanQuery scanQu
...
 
-      closer.register(cursorFromIterable.rhs);
-      final Yielder<Cursor> cursorYielder = Yielders.each(Sequences.simple(ImmutableList.of(cursorFromIterable.lhs)));
+      final Yielder<Cursor> cursorYielder = Yielders.each(Sequences.simple(cursors));
 
       if (cursorYielder.isDone()) {
         // No cursors!
         cursorYielder.close();
-        return ReturnOrAwait.returnObject(Unit.instance());
+        return ReturnOrAwait.returnObject(handedOffSegments);
```
Comment (cryptoe) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/scan/ScanQueryFrameProcessor.java:231]: If there are not handed of segments nulls will be passed. Is that intended ?

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/input/table/TableInputSpecSlicer.java [hunk 3]
```diff
@@ -126,28 +161,67 @@ private Set<DataSegmentWithInterval> getPrunedSegmentSet(final TableInputSpec ta
...
-        retVal.add(new SegmentsInputSlice(tableInputSpec.getDataSource(), descriptors));
+        retVal.add(new SegmentsInputSlice(tableInputSpec.getDataSource(), descriptors, dataServerRequests));
       }
     }
 
     return retVal;
   }
 
-  private static class DataSegmentWithInterval
+  private static List<WeightedInputInstance> createWeightedSegmentSet(List<DataSegmentWithInterval> prunedServedSegments)
```
Comment (cryptoe) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/input/table/TableInputSpecSlicer.java:194]: can you please java doc this method.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/input/table/TableInputSpecSlicer.java [hunk 4]
```diff
@@ -126,28 +161,67 @@ private Set<DataSegmentWithInterval> getPrunedSegmentSet(final TableInputSpec ta
...
+          druidServerMetadataSetEntry.getKey(),
+          ImmutableList.copyOf(druidServerMetadataSetEntry.getValue())
+      );
+      retVal.add(dataServerRequest);
+    }
+
+    return retVal;
+  }
+
+  private interface WeightedInputInstance
```
Comment (cryptoe) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/input/table/TableInputSpecSlicer.java:219]: I think we need docs here as well.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/input/table/TableInputSpecSlicer.java [hunk 5]
```diff
@@ -169,9 +243,40 @@ public RichSegmentDescriptor toRichSegmentDescriptor()
...
+    public DataServerRequest(DruidServerMetadata serverMetadata, List<DataSegmentWithInterval> segments)
+    {
+      this.segments = Preconditions.checkNotNull(segments, "segments");
+      this.serverMetadata = Preconditions.checkNotNull(serverMetadata, "server");
+    }
+
+    @Override
+    public long getWeight()
+    {
+      return segments.size() * DATA_SERVER_WEIGHT_ESTIMATION;
```
Comment (cryptoe) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/input/table/TableInputSpecSlicer.java:272]: We are estimating the number of rows to be 5000 for realtime segments rite ?

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/BaseLeafFrameProcessor.java [hunk 6]
```diff
@@ -110,7 +110,8 @@ protected FrameWriterFactory getFrameWriterFactory()
   }
 
   protected abstract ReturnOrAwait<Unit> runWithSegment(SegmentWithDescriptor segment) throws IOException;
-  protected abstract ReturnOrAwait<Unit> runWithLoadedSegment(SegmentWithDescriptor segment) throws IOException;
+
+  protected abstract ReturnOrAwait<SegmentsInputSlice> runWithDataServerQuery(DataServerQueryHandler dataServerQueryHandler) throws IOException;
```
Comment (cryptoe) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/BaseLeafFrameProcessor.java:114]: Java docs for this method please.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/BaseLeafFrameProcessorFactory.java [hunk 7]
```diff
@@ -146,32 +142,91 @@ public ProcessorsAndChannels<Object, Long> makeProcessors(
...
+            frameContext,
+            factory
+        ),
+        objects -> {
+          if (objects.isEmpty()) {
+            return ProcessorManagers.none();
+          }
+          List<InputSlice> handedOffSegments = new ArrayList<>();
+          for (Object o : objects) {
+            if (o instanceof SegmentsInputSlice) {
```
Comment (cryptoe) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/BaseLeafFrameProcessorFactory.java:213]: O can be null as well.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/BaseLeafFrameProcessorFactory.java [hunk 8]
```diff
@@ -146,32 +142,91 @@ public ProcessorsAndChannels<Object, Long> makeProcessors(
...
+              SegmentsInputSlice slice = (SegmentsInputSlice) o;
+              handedOffSegments.add(slice);
+            }
+          }
+          return new BaseLeafFrameProcessorManager(
```
Comment (cryptoe) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/BaseLeafFrameProcessorFactory.java:218]: I think we are fetching stuff from deep storage here. Lets add a comment for future maintenance.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/ChainedProcessorManager.java [hunk 9]
```diff
@@ -76,22 +91,31 @@ public ListenableFuture<Optional<ProcessorAndCallback<Object>>> next()
...
+
+    //noinspection unchecked
+    return FutureUtils.transformAsync(
+        restFuture,
+        rest -> (ListenableFuture) rest.next()
+    );
   }
 
-  private void onFirstProcessorComplete(final Object firstResult)
+  private synchronized void checkFirstProcessorComplete()
```
Comment (cryptoe) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/ChainedProcessorManager.java:114]: Why is synchronized here ?

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/ChainedProcessorManager.java [hunk 10]
```diff
@@ -24,32 +24,34 @@
...
+import java.util.concurrent.atomic.AtomicInteger;
 import java.util.function.Function;
 
 /**
- * Manager that chains processors: runs {@link #first} first, then based on its result, creates {@link #restFuture}
- * using {@link #restFactory} and runs that next.
+ * Manager that chains processors: runs all processors generated by {@link #first} first, then based on its result,
+ * creates {@link #restFuture} using {@link #restFactory} and runs that next.
  */
 public class ChainedProcessorManager<A, B, R> implements ProcessorManager<Object, R>
```
Comment (cryptoe) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/ChainedProcessorManager.java:42]: @gianm Can you please vet the changes in this class.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/input/table/TableInputSpecSlicer.java (hunk 3 above)
Comment (adarshsanjeev) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/input/table/TableInputSpecSlicer.java:202]: Added

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/input/table/TableInputSpecSlicer.java (hunk 5 above)
Comment (adarshsanjeev) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/input/table/TableInputSpecSlicer.java:282]: Yes, added a comment

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/BaseLeafFrameProcessorFactory.java (hunk 7 above)
Comment (adarshsanjeev) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/BaseLeafFrameProcessorFactory.java]: Added a null check

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/ChainedProcessorManager.java (hunk 9 above)
Comment (cryptoe) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/ChainedProcessorManager.java:114]: Please add documentation regarding which threads need this.

**Review by cryptoe — APPROVED** [review]

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/input/table/SegmentsInputSlice.java [hunk 11]
```diff
@@ -44,15 +44,18 @@ public class SegmentsInputSlice implements InputSlice
...
   private final String dataSource;
   private final List<RichSegmentDescriptor> descriptors;
+  private final List<DataServerRequestDescriptor> servedSegments;
 
   @JsonCreator
   public SegmentsInputSlice(
       @JsonProperty("dataSource") String dataSource,
-      @JsonProperty("segments") List<RichSegmentDescriptor> descriptors
+      @JsonProperty("segments") List<RichSegmentDescriptor> descriptors,
+      @JsonProperty("servedSegments") List<DataServerRequestDescriptor> servedSegments
```
Comment (cryptoe) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/input/table/SegmentsInputSlice.java:53]: This should be marked nullable for rolling upgrade cases.

**Pull request merged**